### PR TITLE
Implement DeployToUpgrade.

### PR DESCRIPTION
**Pull request by phrohdoh**
Will be useful once #7638 is merged.

**Uses**:
- RA2 GI
- TS Nod Artillery (and later GDI Juggernaut) assuming we don't transform them into structures in the same way vanilla TS did
- D2k Thumper

### CODE REVIEW

**Comment by WolfGaming** [comment]
Question though, dont you have to implement IIssueOrders on this since if you do it this way then you will have to have transform on the actor to even send the DeployTransform order.

**Comment by Phrohdoh** [comment]
You do not have to have `Transforms` to send the `DeployTransform` order.
This trait isn't mean to issue an order, only respond to one.

`DeployTransform` is given any time you press the DeployKey (which is F by default).

**Comment by WolfGaming** [comment]
As we talked about on IRC but just for those who dont know of it the IIssueOrder needs to be implemented for the "right click on unit to send order" behaviour and its a more "right" way to do things.

**Comment by Phrohdoh** [comment]
Updated.

**Comment by Phrohdoh** [comment]
Fixed incorrect ordering of using directives.

**Comment by Phrohdoh** [comment]
Made the order always use the deploy cursor.

File: OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+				isUpgraded = false;
+			}
+			else
+			{
+				foreach (var up in info.Upgrades)
+					manager.GrantUpgrade(self, up, this);
+
+				isUpgraded = true;
+			}
+		}
```
Comment (WolfGaming) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: also cant you omit the isUpgraded = false and true there, take the braces and set isUpgraded = !isUpgraded.
aka
```
			if (isUpgraded)
				foreach (var up in info.Upgrades)
					manager.RevokeUpgrade(self, up, this);
			else
				foreach (var up in info.Upgrades)
					manager.GrantUpgrade(self, up, this);

			isUpgraded = !isUpgraded ;
```

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: That is possible.

File: OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+		public object Create(ActorInitializer init) { return new DeployToUpgrade(init.Self, this); }
+	}
+
+	public class DeployToUpgrade : IResolveOrder, IIssueOrder
+	{
+		readonly DeployToUpgradeInfo info;
+		readonly UpgradeManager manager;
+
+		bool isUpgraded;
```
Comment (WolfGaming) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: You need to set an initial value of false to this right?
or will C sharp parse it as false when you do if (null)?

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: `bool` is a ValueType meaning it has a value when you create it.
`bool` defaults to `false`.
A bool cannot be compared to `null` as `null` is for ReferenceTypes.

**Comment by WolfGaming** [comment]
Then we should probably change the F key to send only one order named Deploy which everyone checks if we want to do it that way.

Now if we had a place where we could display icons that activate abilities and we use them to parse the hotkeys then we could remove this hacky approach.

**Comment by Phrohdoh** [comment]
That is completely out of the scope of this PR.

**Comment by WolfGaming** [comment]
I know, just throwing it out there.
I just dont like a trait resolving orders it has no buisness to resolve, this will need to be fixed if we ever implement what I said above.

File: OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs (hunk 2 above)
Comment (WolfGaming) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: Sometimes I dont think do I :/

**Comment by Phrohdoh** [comment]
Updated.

**Comment by Phrohdoh** [comment]
Changed `DeployOrderTargeter` to use a `string cursor` param instead of falling onto either `deploy` or `deploy-blocked`.

**Comment by Phrohdoh** [comment]
I've added the `Help wanted` label because I just thought of an abuse situation with this (and/or a bug):

If you use this and Transforms the actor will be granted the upgrade as expect then will become another actor.

When undeploying the new actor may not have this trait (and even if it does, it will be a separate instance of `DeployToUpgrade`) so the upgrade will not be revoked and `isUpgraded` will not be unset properly.

I do not know how to solve this.

**Comment by pchote** [comment]
> If you use this and Transforms the actor will be granted the upgrade as expect then will become another actor.

Simple: don't do this. It is a pretty bogus thing for somebody to do.

**Comment by Phrohdoh** [comment]
That is a completely bogus answer in my opinion, but okay.

**Comment by pchote** [comment]
The use case for Transforms is situations where you don't want much of the earlier state to persist. The purpose of this trait is to modify state without doing a transform. If you want a state change to occur at the same time as a transformation, then you can just define that in the transformed actor type.  I can't think of any hypothetical cases where you would need to use the two together.

**Comment by Mailaender** [comment]
Needs a rebase. Is https://github.com/OpenRA/OpenRA/pull/7638 a hard dependency of this?

**Comment by Phrohdoh** [comment]
@Mailaender not it is not a dependency at all.

**Comment by Phrohdoh** [comment]
Rebased.

File: OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		{
+			if (order.OrderID == "DeployUpgrade")
+				return new Order(order.OrderID, self, queued);
+
+			return null;
+		}
+
+		public void ResolveOrder(Actor self, Order order)
+		{
+			// DeployTransform is necessary to use the DeployKey (F)
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: Simply add it at https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Widgets/WorldCommandWidget.cs#L135 instead of hiding the hack here.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: This has been requested before and my answer was no, that is adding more to a hack that already exist, which I won't do.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: Your hack at this position is no better just more sneaky.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: ¯\_(ツ)_/¯

Comment (penev92) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: I don't get it. What does this have to do with the DeployKey?

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: Keeping all the hacks in the same place means that they can all be fixed at the same time.  Hiding sneaky little hacks all across the code base is a great way to make future developers swear at you.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: I also don't believe we should add more hacks and say "one day" so I can close this if desired.

Comment (penev92) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: I really don't think closing this is the desired thing.
And I still don't get what the `DeployKey` has to do with anything, or what "DeployTransform" has to do with this trait.
As we discussed on IRC, this trait is here precisely to do the stuff Transforms does, just better. Not to be used together with it.
Or am I missing something in the implementation?

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: This is absolutely **not** do to what `Transforms` does.
`Transforms` is used when you don't care about keeping the current state/data.
This is used when you *do* care and want it to persist (on undeploy).

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: If you're not prepared to work with reviewers towards getting a pr merged then you should just close this to avoid wasting any more of our time.

Comment (penev92) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: OK, so the input handler works under the assumption that every actor has no more than one Stop/Scatter/Deploy/etc. action. Which, I guess, is fine. We really shouldn't have both this and Transforms on a single avtor though.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: The input handler is not making any assumptions (in general).  The "problem" is that we have not built out proper infrastructure for exposing orders to the UI (via an order palette or hotkeys) and so they are currently hardcoded in the world viewport widget.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: >If you're not prepared to work with reviewers towards getting a pr merged then you should just close this to avoid wasting any more of our time.

I'm not really sure what you're issue is with me not wanting to pile on another hack, but okay if you insist.

**Comment by Phrohdoh** [comment]
Moved the hack to another place.

File: OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs (hunk 3 above)
Comment (WolfGaming) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: It doesnt matter how you do it, in the DeployToUpgrade class or the world command widget, both are hacks, the one in DeployToUpgrade is just more hidden and if we ever implement something to send those orders it is better to have them in one place rather then in each trait.

File: OpenRA.Mods.Common/Orders/DeployOrderTargeter.cs [hunk 4]
```diff
@@ -16,18 +16,18 @@ namespace OpenRA.Mods.Common.Orders
...
+		readonly string cursor;
 
 		public DeployOrderTargeter(string order, int priority)
-			: this(order, priority, () => true)
+			: this(order, priority, "deploy")
 		{
 		}
 
-		public DeployOrderTargeter(string order, int priority, Func<bool> useDeployCursor)
+		public DeployOrderTargeter(string order, int priority, string cursor)
```
Comment (penev92) [inline OpenRA.Mods.Common/Orders/DeployOrderTargeter.cs]: I don't get this change. Could you explain?

File: OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class DeployToUpgradeInfo : ITraitInfo, Requires<UpgradeManagerInfo>
+	{
+		[Desc("The upgrades to grant when deploying and revoke when undeploying.")]
+		public readonly string[] Upgrades = { };
+
+		[Desc("Cursor to display when deploying the actor.")]
+		public readonly string Cursor = "deploy";
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: What's the use-case for this being configurable?

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: It also documents it so I think exposing this with a set default is always a good idea.

**Comment by penev92** [comment]
I don't see this considering whether the actor **can** deploy.
I'd like to see a check for that, along with a field in the Info class for the disabled cursor, if the cursors need to be configurable.

**Comment by Phrohdoh** [comment]
There is no way to check wherever or not the actor can deploy, unless you want to implement the interface we discussed. I can add a disabled cursor though.

**Comment by penev92** [comment]
Sorry, I don't recall what interface we've discussed.
If that's not easy to do at the moment, then consider this:
The main thing about "can I deploy here" is "can I deploy **on this terrain**". GIs can deploy anywhere, any time, but thumpers can only deploy on sand, and artilleries should be able to deploy only on flat land. Also they may or may not be able to deploy on iced-over rivers, I can't remember that (but it makes no sense to be able to).

All the other possible use-cases for this that I can think of care about:
**1.** The terrain type.
**2.** The terrain inclination.
**3.** Free space around them.

The immediate 3 (4) use-cases (GI, thumper, Tick Tank, arty) care about the first two, so I'd like to see the trait's Info class have fields for valid deployment terrains and whether the actor is allowed to deploy on a slope. The free space is probably harder to do and not as urgent, I feel.

**Comment by Phrohdoh** [comment]
You are right about the slope concern, but my issue with that is the interface (IPreventsDeploy) needs to go on a trait which then checks slopes, terrain types, etc.

I don't know of a good logical layout for a trait to check all of the necessary conditions.

We could take the approach that the upgrade a manger does (again, on the interface) which is passing the `source` object.

Needs discussion.

File: OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs (hunk 5 above)
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: Configurability > hardcoded values.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Upgrades/DeployToUpgrade.cs]: RA2 includes artwork for multiple different deploy cursor types (IIRC) that we may want to use.

**Comment by Phrohdoh** [comment]
Unfortunately I messed this up fooling with git so I must open another PR.
Apologies.

**Comment by Phrohdoh** [comment]
Please refer to #7865.
